### PR TITLE
pluto: improve cluster DNS IP generation

### PR DESCRIPTION
**Pull request by jpculp**
**Description of changes:**

Rather than assuming IPv4, match the `ipFamily` in the EKS cluster's `kubernetesNetworkConfig`.

**Testing done:**

- Launched Bottlerocket AMI attached to IPv6 cluster.
- Launched Bottlerocket AMI attached to IPv4 cluster.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

**Review by Issacwww — APPROVED** [review]

File: sources/api/pluto/src/main.rs
Comment (cbgbt) [inline sources/api/pluto/src/main.rs:274]: The nested if let statements in the match arms make this a bit dense. Consider flattening the logic using and_then:

```suggestion
    // Derive cluster-dns-ip from service CIDR based on IP family
    match config.ip_family {
        Some(IpFamily::Ipv6) => config
            .service_ipv6_cidr
            .and_then(|cidr| get_dns_from_ipv6_cidr(&cidr).ok()),
        _ => config
            .service_ipv4_cidr
            .and_then(|cidr| get_dns_from_ipv4_cidr(&cidr).ok()),
    }
    .map(Some)
    .unwrap_or(Ok(None))
```

File: sources/api/pluto/src/main.rs [hunk 1]
```diff
@@ -557,6 +587,21 @@ mod test {
         assert!(result.is_err());
     }
 
+    #[test]
```
Comment (cbgbt) [inline sources/api/pluto/src/main.rs]: Given the issue with compressed IPv6 addresses, can you add more test cases? I really like the [test-case crate](https://crates.io/crates/test_case) because it lets you write test case data in a way that still preserves pleasant output from `cargo test`. We use this dependency elsewhere in Bottlerocket.

```rust
#[test_case("fd00::/108", "fd00::a" ; "compressed ULA - common in EKS")]
#[test_case("2001:db8::/108", "2001:db8::a" ; "compressed middle")]
#[test_case("2001:db8:1234::/108", "2001:db8:1234::a" ; "partially expanded")]
#[test_case("2001:db8:1234:5678::/108", "2001:db8:1234:5678::a" ; "fully expanded")]
#[test_case("fe80::/108", "fe80::a" ; "link local")]
fn test_get_dns_from_ipv6_cidr_variants(input: &str, expected: &str) {
    let actual = get_dns_from_ipv6_cidr(input).unwrap();
    assert_eq!(expected, actual);
}

#[test_case("not-an-ipv6" ; "no slash")]
#[test_case("" ; "empty string")]
fn test_get_dns_from_ipv6_cidr_errors(input: &str) {
    let result = get_dns_from_ipv6_cidr(input);
    assert!(result.is_err());
}
```

File: sources/api/pluto/src/main.rs [hunk 2]
```diff
@@ -296,6 +308,24 @@ fn get_dns_from_ipv4_cidr(cidr: &str) -> Result<String> {
     split[3] = "10";
     Ok(split.join("."))
 }
+/// Replicates [this] logic from the EKS AMI:
+///
+/// ```sh
+/// DNS_CLUSTER_IP=$(awk -F/ '{print $1}' <<< $SERVICE_IPV6_CIDR)a
+/// ```
+/// [this]: https://github.com/awslabs/amazon-eks-ami/blob/0f5c129/templates/al2/runtime/bootstrap.sh#L463
+fn get_dns_from_ipv6_cidr(cidr: &str) -> Result<String> {
```
Comment (cbgbt) [inline sources/api/pluto/src/main.rs:316]: This won't work with compressed IPv6 notation. AWS EKS returns compressed ULA addresses like `fd00::/108`, which would only have 3 components when split by `:`.

Looking at the [referenced EKS AMI code](https://github.com/awslabs/amazon-eks-ami/blob/0f5c129/templates/al2/runtime/bootstrap.sh#L463), the bash implementation is:

```bash
DNS_CLUSTER_IP=$(awk -F/ '{print $1}' <<< $SERVICE_IPV6_CIDR)a
```

This splits on `/`, takes the first part, and appends `a`, which will work with compressed addresses regardless of the number of segments they have.

We could use a proper IPv6 parser here, since we have the benefit of not using bash:

```rust
use std::net::Ipv6Addr;

fn get_dns_from_ipv6_cidr(cidr: &str) -> Result<String> {
    // Extract the address part before the slash
    let addr_str = cidr.split('/').next().context(error::CidrParseSnafu {
        cidr,
        reason: "missing address component",
    })?;

    let base_addr: Ipv6Addr = addr_str.parse().context(error::BadIpSnafu {
        ip: addr_str,
    })?;

    // Set the last segment to 0xa (10 in hex)
    let mut segments = base_addr.segments();
    segments[7] = 0xa;

    Ok(Ipv6Addr::from(segments).to_string())
}
```

**Comment by jpculp** [comment]
Addressed feedback from @cbgbt and re-ran tests.

File: sources/api/pluto/src/main.rs [hunk 3]
```diff
@@ -609,6 +655,43 @@ mod test {
...
+        }
+
+        if let Some(cluster_name) = cluster_name {
+            settings_view_set!(settings.kubernetes.cluster_name = cluster_name.try_into().unwrap());
+        }
+
+        // Since we can't easily mock the EKS service in unit tests, we test the logic paths
+        // The function will attempt to call EKS when both region and cluster_name are present,
+        // but will fail gracefully in test environment and return Ok(None)
+        let result = get_eks_network_config(&settings).await;
```
Comment (cbgbt) [inline sources/api/pluto/src/main.rs]: We shouldn't make network API calls in tests, even if we know they will fail. This limitation also makes it impossible for us to test the positive case, and only allows us to test the negative.

It's not easy, but we should either:
* Modify the logic to make mocking possible (Rust's `mockall` crate is excellent, but you will need to introduce traits)
* Find a different way to isolate the compute logic from the IO, and then only test the logic

Comment (cbgbt) [reply sources/api/pluto/src/main.rs]: Here's an example of how you could improve the unit tests: https://gist.github.com/cbgbt/770870b6f292f9e60b21e6913ebcf259

**Comment by jpculp** [comment]
Moved DNS block into its own function and simplified testing based on @cbgbt's suggestion.

**Review by cbgbt — APPROVED** [review]

**Review by piyush-jena — APPROVED** [review]

**Review by bcressey — APPROVED** [review]
